com.microsoft.mrtk.graphicstools.unity/Runtime/Experimental/AmbientOcclusion/AmbientOcclusion.cs

```diff
@@ -0,0 +1,394 @@
+                _vertexes[vi] = transform.TransformPoint(_vertexes[vi]);
+                _normals[vi] = transform.TransformVector(_normals[vi]);
+
+                _hitsInHemisphere.Clear();
+
+                Vector3 averageDir = Vector3.zero;
+
+                var origin = _vertexes[vi] + _normals[vi].normalized * kOriginNormalOffset;
+
+                for (int ni = 0; ni < SamplesPerVertex; ni++)
+                {
+                    var sampleDirection = RandomSampleAboveHemisphere(_normals[vi]);
+
+                    // Visualization: Save samples for gizmo
+                    if (vi == ReferenceVertexIndex && _showSamples)
+                    {
+                        _referenceVertexSamples.Add(sampleDirection);
+                    }
+
+                    if (Physics.Raycast(origin, sampleDirection, out hit, MaxSampleDistance))
+                    {
+                        _hitsInHemisphere.Add(hit);
+
+                        // Visualization: Save hits for gizmo
+                        if (vi == ReferenceVertexIndex)
+                        {
+                            _referenceVertexHits.Add(hit);
+                        }
+                    }
+                    else
+                    {
+                        averageDir += sampleDirection;
+                    }
+                }
+                var avgN = averageDir.normalized;
+                _bentNormalsAo[vi] = new Vector4(avgN.x,
+                                                 avgN.y,
+                                                 avgN.z,
+                                                 1 - ((float)_hitsInHemisphere.Count / SamplesPerVertex));
+            }
+
+            SourceMesh.SetUVs(kUvChannel, _bentNormalsAo);
+
+            UnityEngine.Debug.Log($"{nameof(GatherSamples)} elapsed-ms={watch.ElapsedMilliseconds} vertex-count={SourceMesh.vertexCount} rays-per-vertex={SamplesPerVertex}.");
+        }
+
+        private void SaveOriginalColors()
+        {
+            UnityEngine.Debug.Log(nameof(SaveOriginalColors));
+            if (SourceMesh.colors.Length == SourceMesh.vertexCount)
+            {
+                _originalHasColors = true;
+                _originalColors = new Color[SourceMesh.vertexCount];
+                Array.Copy(SourceMesh.colors, _originalColors, SourceMesh.vertexCount);
+                SourceMesh.GetUVs(kUvChannel, _originalUv4);
+            }
+        }
+
+        private void RestoreColors()
+        {
+            // XXX see if we can reuse old data without re-sample
+            UnityEngine.Debug.Log(nameof(RestoreColors));
+        }
+
+        public void RestoreOriginalColors()
+        {
+            if (_originalColors == null)
+            {
+                SaveOriginalColors();
+            }
+
+            var colors = new Color[_originalColors.Length];
+
+            if (_originalHasColors)
+            {
+                for (int i = 0; i < _originalColors.Length; i++)
```

Comment:
Do you need this loop or can you just assign _originalColors to SourceMesh.colors?

Reply:
Fixed.